internal/controller/controllers/image_controller.go

```diff
@@ -17,29 +17,127 @@ limitations under the License.
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
 
+	"github.com/go-openapi/strfmt"
+	"github.com/openshift/assisted-service/internal/bminventory"
+	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/restapi/operations/installer"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-
-	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
 )
 
 // ImageReconciler reconciles a Image object
 type ImageReconciler struct {
 	client.Client
-	Log    logrus.FieldLogger
-	Scheme *runtime.Scheme
+	Log       logrus.FieldLogger
+	Scheme    *runtime.Scheme
+	Installer bminventory.InstallerInternals
 }
 
 // +kubebuilder:rbac:groups=adi.io.my.domain,resources=images,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=adi.io.my.domain,resources=images/status,verbs=get;update;patch
 
 func (r *ImageReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// your logic here
+	ctx := context.Background()
+
+	image := &adiiov1alpha1.Image{}
+	if err := r.Get(ctx, req.NamespacedName, image); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.create(ctx, image)
+}
+
+func (r *ImageReconciler) create(ctx context.Context, image *adiiov1alpha1.Image) (ctrl.Result, error) {
+	state := adiiov1alpha1.ImageStateFailedToCreate
+
+	cluster, clusterRefErr := r.getClusterByRef(ctx, image.Spec.ClusterRef)
+	if clusterRefErr != nil {
+		return r.updateStatusAndReturnResult(ctx, image, nil, state, clusterRefErr)
+	}
+
+	var imageInfo *models.ImageInfo
+	updatedCluster, inventoryErr := r.Installer.GenerateClusterISOInternal(
+		ctx,
+		installer.GenerateClusterISOParams{
+			ClusterID: strfmt.UUID(cluster.Status.ID),
+			ImageCreateParams: &models.ImageCreateParams{
+				SSHPublicKey:    image.Spec.SSHPublicKey,
+				StaticIpsConfig: image.Spec.StaticIpConfiguration,
+			},
+		})
+	if inventoryErr == nil {
+		imageInfo = updatedCluster.ImageInfo
+		state = adiiov1alpha1.ImageStateCreated
+	}
+
+	return r.updateStatusAndReturnResult(ctx, image, imageInfo, state, inventoryErr)
+}
+
+func (r *ImageReconciler) getClusterByRef(
+	ctx context.Context,
+	ref *corev1.ObjectReference) (*adiiov1alpha1.Cluster, error) {
+	key := types.NamespacedName{
+		Name:      ref.Name,
+		Namespace: ref.Namespace,
+	}
+	cluster := &adiiov1alpha1.Cluster{}
+	if err := r.Get(ctx, key, cluster); err != nil {
+		return nil, newKubeAPIError(
+			errors.Wrapf(
+				err,
+				fmt.Sprintf(
+					"failed to find cluster with name %s in namespace %s",
+					ref.Name, ref.Namespace)),
+			k8serrors.IsNotFound(err))
```

Comment:
You don't know if it's not found or not it will just return boolean so what will you gain by just  adding boot to the error?

Reply:
the returned value of `k8serrors.IsNotFound(err)`goes to the IsClientError field of the KubeAPIError which will eventually determine whether to requeue the request.